jurity/recommenders/rank_estimation.py

```diff
@@ -0,0 +1,47 @@
+from typing import NamedTuple
+
+import numpy as np
+import pandas as pd
+
+from jurity.utils import Constants, get_sorted_clicks
+
+
+class RankEstimation:
+    def __init__(self, n_items: int = 1, n_sampled: int = 1):
+        """ Unbiased Rank Estimation
+
+        Performs a correction of ranking metrics for use cases where the negative responses were sampled.
+        This adjusts the metric for when if the full set of items were used,
+        items in the sampled top k would not be ranked as high.
+
+        .. math::
+            URE = metric \\times (1 + \\frac{(n_sampled -1)(predicted_rank - 1)}{n_items})
+
+        Parameters
+        ----------
+        n_items: int
+            The total number of items available for recommendation.
+        n_sampled: int
+            The number of items sampled for the response matrix.
+        """
+
+        self.n_items = n_items
+        self.n_sampled = n_sampled
+
+    def get_correction(self, predicted_results: pd.DataFrame):
```

Comment:
I know I'm not supposed to get into implementation questions, but how will this be used. Will the correction be applied to the input `predicted_results` and then the ranking metric is calculated as per usual?

Reply:
Yes, it calculates the correction for each individual prediction (the r in the equation in the docstring is the predicted rank of the item)

Reply:
Note to ourselves: these corrections help when evaluating on _test data_, when I train on sampled train data and then test on real unsampled data, nothing can help me, right? :D

Reply:
in theory the greater reliability of doubly robust evaluation should help with that ;)